sdks/python/apache_beam/runners/interactive/display/pcoll_visualization.py

```diff
@@ -0,0 +1,269 @@
+    assert isinstance(pcoll, pvalue.PCollection), (
+        'pcoll should be apache_beam.pvalue.PCollection')
+    self._pcoll = pcoll
+    # This allows us to access cache key and other meta data about the pipeline
+    # whether it's the pipeline defined in user code or a copy of that pipeline.
+    # Thus, this module doesn't need any other user input but the PCollection
+    # variable to be visualized. It then automatically figures out the pipeline
+    # definition, materialized data and the pipeline result for the execution
+    # even if the user never assigned or waited the result explicitly.
+    # With only the constructor of PipelineInstrument, any interactivity related
+    # pre-process or instrument is not triggered for performance concerns.
+    self._pin = instr.PipelineInstrument(pcoll.pipeline)
+    self._cache_key = self._pin.cache_key(self._pcoll)
+    self._dive_display_id = 'facets_dive_{}_{}'.format(self._cache_key,
+                                                       id(self))
+    self._overview_display_id = 'facets_overview_{}_{}'.format(self._cache_key,
+                                                               id(self))
+    self._df_display_id = 'df_{}_{}'.format(self._cache_key, id(self))
+
+  def display_facets(self, updating_pv=None):
+    """Displays the visualization through IPython.
+
+    Args:
+      updating_pv: A PCollectionVisualization object. When provided, the
+        display_id of each visualization part will inherit from the initial
+        display of updating_pv and only update that visualization web element
+        instead of creating new ones.
+
+    The visualization has 3 parts: facets-dive, facets-overview and paginated
+    data table. Each part is assigned an auto-generated unique display id
+    (the uniqueness is guaranteed throughout the lifespan of the PCollection
+    variable).
+    """
+    # Ensures that dive, overview and table render the same data because the
+    # materialized PCollection data might being updated continuously.
+    data = self._to_dataframe()
+    if updating_pv:
+      self._display_dive(data, updating_pv._dive_display_id)
+      self._display_overview(data, updating_pv._overview_display_id)
+      self._display_dataframe(data, updating_pv._df_display_id)
+    else:
+      self._display_dive(data)
+      self._display_overview(data)
+      self._display_dataframe(data)
+
+  def _display_dive(self, data, update=None):
+    sprite_size = 32 if len(data.index) > 50000 else 64
+    jsonstr = data.to_json(orient='records')
+    if update:
+      script = _DIVE_SCRIPT_TEMPLATE.format(display_id=update,
+                                            jsonstr=jsonstr)
+      display_javascript(Javascript(script))
+    else:
+      html = _DIVE_HTML_TEMPLATE.format(display_id=self._dive_display_id,
+                                        jsonstr=jsonstr,
+                                        sprite_size=sprite_size)
+      display(HTML(html))
```

Comment:
For ipython terminal, the users would see something similar to below:
```
<IPython.core.display.HTML object>
<IPython.core.display.HTML object>
<IPython.core.display.HTML object>
```
I think the `show`(`visualization`) here heavily depends on HTML/JS/CSS to make sense (because it tries to load the whole PCollection cache into memory for rendering), so if the kernel is connected to a frontend that supports these, they should all support the visualization.

We can add a default pretty print that works similar to existing `print(result.get_list(pcoll))` and it's nothing close to what current `visualization()` does. However, the kernel doesn't know whether it is connected to a frontend / what kinds of frontends, so it's either always pretty-print or not doing it at all.
I would propose limit the work here for the scenarios when the kernel is actually connected to a frontend. And data visualization in a plain ipython terminal is not what we want to achieve here since one cannot interact with the rendering in a terminal easily (with mouse and etc.).

Reply:
I see. I don't want to block on this, so LGTM. But robert had also mentioned that he wanted this to be useful outside of a Jupyter notebook. Also the Ipython terminal. We can discuss this separately.

Reply:
I mean I can add something to print a dataframe-printing-like output in terminal that samples some of the data and use some "hacky" way (details can be found in this stackoverflow [discussion](https://stackoverflow.com/questions/15411967/how-can-i-check-if-code-is-executed-in-the-ipython-notebook)) to determine if the kernel is connected to an IPython notebook.

I'll send an additional commit in this PR to implement that.

But we are not going to provide other GUI support in terminal for now.